Index: src/svncpp/context.cpp
===================================================================
--- src/svncpp/context.cpp	(revision 7023)
+++ src/svncpp/context.cpp	(working copy)
@@ -31,6 +31,11 @@
   struct Context::Data
   {
   public:
+    /** The usage of Apr makes sure Apr is initialized
+     * before any use of apr functions.
+     */
+    Apr apr;
+
     ContextListener * listener;
     bool logIsSet;
     int promptCounter;
@@ -41,9 +46,6 @@
     std::string logMessage;
     std::string configDir;
 
-    /** The usage of Apr makes sure Apr is initialized */
-    Apr apr;
-
     /**
      * translate native c-string to utf8 
      */
